Add more NRT examples

BillWagner opened the pull request:
Fixes #1214

Add examples as noted in the referenced comments.

BillWagner committed:
Add more NRT examples

Fixes #1214

Add examples as noted in the referenced comments.

BillWagner committed:
fix grammar note

BillWagner committed:
Fix examples

BillWagner committed:
update expected exception

Nigel-Ecma commented on standard/types.md:
````diff
@@ -936,15 +936,64 @@ The compiler can update the null state of a variable as part of its analysis.
+>         _ = s.Length; // warning
+>     }
+> }
+> ```
````
Should the next line (964) markdownlint-disable be deleted? It isn't paired with an enable.

Nigel-Ecma commented on standard/types.md:
````diff
@@ -936,15 +936,64 @@ The compiler can update the null state of a variable as part of its analysis.
+> ```
 >
-> <!-- Example: {template:"standalone-console", name:"NullPropertyAnalysis"} -->
+> In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*
````
Example above has been replaced, this para describes the old version.

Nigel-Ecma reviewed (changes requested):
There is a left-over para, or one that was going to be replaced.

I think “The compiler” here should probably always be ”A compiler” – there are a few of the latter already.

BillWagner committed:
Replace "the compiler" with "a compiler", part 1.

Replace instances of "the compiler" with "a compiler" as it relates to what actions a compiler might take while performing static nullable analysis.

BillWagner commented:
@Nigel-Ecma @RexJaeschke 

> I think “The compiler” here should probably always be ”A compiler” – there are a few of the latter already.

I did replace instances of "the compiler" as it relates to this PR with "a compiler".  A search in the standard uncovers 68 other instances of "the compiler". Some should not be changed. They are part of a longer phrase like "the compiler generated backing field" for properties and field like events.

Should I change others where the language is similar to those Nigel mentioned in this PR? Should I add it to this PR, or create a separate smaller PR?

RexJaeschke commented:
@BillWagner Yes, please change them, as appropriate. If there are occurrences in other topics then I think a separate PR would be best, making it a general editorial fix.

Nigel-Ecma commented:
@BillWagner & @RexJaeschke – I agree with Rex to change the ones you see.

It is easy to go down a rabbit hole here; I started as Bill did and saw each “the” needs to be examined – e.g. as in the backing field example “the” refers to the field and not the compiler. From there I started looking into “automatically” – as in “the compiler automatically…” – and these often should at least be something like “a compiler shall/must/may” – this is a spec not a description after all. Once we get to these we need to decide what is being *required* etc. Rabbit hole! This language I expect in many cases dates from the descriptive material from which the 1st edition was written, it will be a long road and incremental is the way to go.

jskeet commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+>     // s is not null:
+>     int length = s.Length;
+>
+>     _ = s = null; // Null check by testing equality. The null state of o is maybe null
```
Is this meant to be `s == null`? At the moment this is an assignment...

jskeet replied:
(And what is "o" referred to in the comment? I'm guessing s used to be o - needs to be changed below in multiple places.)

jskeet commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+>     _ = s?.Length; // The ?. is a null check and changes the null state of o to maybe null
+>     if (s.Length > 4) // Warning. Changes null state of o to not null
+>     {
+>        _ = s?[4]; //is a null check and changes the null state of o to maybe null
```
We're missing a noun at the start of the sentence, and a space after `//`

jskeet commented on standard/types.md:
````diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+> }
+> ```
+
+Both auto-property and field like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.
````
```suggestion
Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.
```

jskeet replied:
(I'd personally prefer to hyphenate compiler-generated everywhere as well, but I'm not doing that now. Field-like events are hyphenated consistently though, I believe.)

jskeet commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+
+Both auto-property and field like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.
+
+> *Example*: A compiler can determine that writing an auto-property or field like event writes the corresponding compiler generated backing field. The null state of the property matches that of the backing field.
```
```suggestion
> *Example*: A compiler can determine that writing an auto-property or field-like event writes the corresponding compiler generated backing field. The null state of the property matches that of the backing field.
```

jskeet commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+>     int length = s.Length;
+>
+>     _ = s = null; // Null check by testing equality. The null state of o is maybe null
+>     length = s.Length; // warning, and changes the null state of o to not null
```
```suggestion
>     length = s.Length; // Warning, and changes the null state of o to not null
```

jskeet commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+>     static void Main()
+>     {
+>         var t = new Test();
+>         int len = t.P.Length; // No warning. null state is not null
```
```suggestion
>         int len = t.P.Length; // No warning. Null state is not null.
```

jskeet commented on standard/types.md:
````diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+> }
+> ```
+>
+> In the previous example, the constructor doesn't set `P` to a not null value, and a compiler issues a warning. There's no warning when the `P` property is accessed, because the type of the property is a non nullable reference type. *end example*
````
> and a compiler issues a warning

Issues, or may issue?

jskeet commented on standard/types.md:
```diff
@@ -976,4 +1024,26 @@ Later in the method, the code checks that `s` is not a null reference. The null-
 >
 > In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*
 
+A compiler may use any expression that dereferences a variable, property, or event to set the null state to not null. If it were null, the dereference expression would have thrown a `NullReferenceException`:
```
Do we want to add a note about extension methods, where calling an extension method *looks* like it dereferences a variable, but doesn't actually do so?

BillWagner commented on standard/types.md:
```diff
@@ -936,15 +936,63 @@ The compiler can update the null state of a variable as part of its analysis.
+>     // s is not null:
+>     int length = s.Length;
+>
+>     _ = s = null; // Null check by testing equality. The null state of o is maybe null
```
```suggestion
>     _ = s == null; // Null check by testing equality. The null state of s is maybe null
```

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
Respond to reviews

BillWagner committed:
respond to feedback 2.

BillWagner commented:
> Yes, please change them, as appropriate. If there are occurrences in other topics then I think a separate PR would be best, making it a general editorial fix.

Added in #1126

BillWagner committed:
fix expected warnings.

jskeet commented on standard/types.md:
````diff
@@ -912,39 +912,83 @@ A diagnostic can be produced when a variable ([§9.2.1](variables.md#921-general
+> }
+> ```
+
+Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.
````
```suggestion
Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state analysis may assume that assignment to the event or property is an assignment to a compiler generated backing field.
```

jskeet replied:
(Or change "assume that" to "treat".)

jskeet commented on standard/types.md:
```diff
@@ -968,12 +1012,34 @@ Later in the method, the code checks that `s` is not a null reference. The null-
 >         if (t.DisappearingProperty != null)
 >         {
->             int len = t.DisappearingProperty.Length;
+>             int len = t.DisappearingProperty.Length; // No warning. A compiler can assume state
```
I suspect the second sentence *is* as intended, but sounds incomplete to me.

BillWagner replied:
reworded that sentence.

jskeet commented on standard/types.md:
```diff
@@ -968,12 +1012,34 @@ Later in the method, the code checks that `s` is not a null reference. The null-
+>     {
+>         _ = child.child.child; // Warning. Dereference possible null value
+>         var greatGrandChild = child.child.child; // No warning. 
+>
```
Remove blank line?

BillWagner committed:
Update standard/types.md

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
respond to feedback.

jnm2 commented on standard/types.md:
````diff
@@ -912,39 +912,83 @@ A diagnostic can be produced when a variable ([§9.2.1](variables.md#921-general
+> }
+> ```
+
+Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state analysis may infer that assignment to the event or property is an assignment to a compiler generated backing field.
````
Uber nitpick :D
```suggestion
Both auto-property and field-like event declarations make use of a compiler-generated backing field. Null state analysis may infer that assignment to the event or property is an assignment to a compiler generated backing field.
```

jnm2 replied:
Also "auto property" instead of "auto-property" but that one is likely a separate item if it's even palatable to consider.

BillWagner committed:
Update standard/types.md

Co-authored-by: Joseph Musser <me@jnm2.com>

jskeet commented:
Approved 2024-12-11 call.
